models/events.go

```diff
@@ -0,0 +1,121 @@
+package models
+
+import (
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"golang.org/x/exp/slices"
+)
+
+var evmLocation common.Location = flow.EVMLocation{}
+
+var BlockExecutedEventType = evmLocation.TypeID(
+	nil,
+	string(types.EventTypeBlockExecuted),
+)
+
+var TransactionExecutedEventType = evmLocation.TypeID(
+	nil,
+	string(types.EventTypeTransactionExecuted),
+)
+
+// isBlockExecutedEvent checks whether the given event contains block executed data.
+func isBlockExecutedEvent(event cadence.Event) bool {
+	if event.EventType == nil {
+		return false
+	}
+	return common.TypeID(event.EventType.ID()) == BlockExecutedEventType
+}
+
+// isTransactionExecutedEvent checks whether the given event contains transaction executed data.
+func isTransactionExecutedEvent(event cadence.Event) bool {
+	if event.EventType == nil {
+		return false
+	}
+	return common.TypeID(event.EventType.ID()) == TransactionExecutedEventType
+}
+
+type CadenceEvents struct {
+	events flow.BlockEvents
+}
+
+func NewCadenceEvents(events flow.BlockEvents) *CadenceEvents {
+	return &CadenceEvents{events: events}
+}
+
+// Blocks finds the block evm events and decodes it into the blocks slice,
+// if no block events are found nil slice is returned.
+//
+// Return values:
+// blocks, nil - if blocks are found
+// nil, nil - if no block are found
+// nil, err - unexpected error
+func (c *CadenceEvents) Blocks() ([]*types.Block, error) {
+	blocks := make([]*types.Block, 0)
+	for _, e := range c.events.Events {
+		if isBlockExecutedEvent(e.Value) {
+			block, err := decodeBlock(e.Value)
+			if err != nil {
+				return nil, err
+			}
+			blocks = append(blocks, block)
+		}
+	}
+
+	// make sure order of heights is ordered, this is safety check
+	slices.SortFunc(blocks, func(a, b *types.Block) int {
+		return int(a.Height - b.Height)
+	})
+
+	return blocks, nil
+}
+
+// Transactions finds all the transactions evm events and decodes them into transaction slice,
+// if no transactions is found nil is returned.
+//
+// Return values:
```

Comment:
Maybe we should also document the return value of receipts, as the function returns 3 values.